wallet-new/src/Cardano/Wallet/Kernel/DB/TxMeta/Types.hs

```diff
@@ -184,12 +183,17 @@ instance Buildable TxMetaStorageError where
     build storageErr = bprint shown storageErr
 
 -- | Generates 'NonEmpty' collections which do not contain duplicates.
--- Limit the size to @size@ elements.
+-- Limit the size to @size@ elements. @size@ should be > 0.
 uniqueElements :: (Arbitrary a, Ord a) => Int -> Gen (NonEmpty a)
+uniqueElements 0 = error "should have size > 0"
 uniqueElements size = do
-    noDupes <- suchThat arbitrary (\s -> length s >= size)
-    let (e, es) = Set.deleteFindMax noDupes
-    return $ e :| List.take size (Set.toList es)
+    (NonEmpty.fromList . Set.toList) <$> (go mempty)
+    where
+      go st = if (Set.size st == size)
+        then return st
+        else do
+          a <- arbitrary
+          go (Set.insert a st)
```

Comment:
This is now OK, but note how this is probably not solving anything. You have replaced `suchThat` with a recursive algorithm where the terminating condition is subordinate to the number of (unique) elements in the set. It can take a very long time for this set to reach the requested `size`.

This definitely brings us closer to a `suchThat` free code, but I would dare to guess that in terms of asymptotic complexity this version is no faster than the previous one.

I suspect the real solution might be to reuse a pool of unique transactions pre-generated statically ahead-of-time, and reuse this throughout all the tests. This is nothing worth preventing this PR from being merged, just something to be aware of 😉

Reply:
@adinapoli-iohk  I have a very naive understanding of QuickChek, but I ran some benchmarks and I saw some really good improvements with this fix. These are the times I got from the generator when creating Inputs sized in [10, 20 .. 500]:
before: ["0.000366s","0.000164s","0.00074s","0.002048s","0.002395s","0.007534s","0.027384s","0.050889s","0.086268s","0.104258s","0.168888s","0.255049s","0.255296s","0.334224s","0.43517s","0.527902s","0.633869s","0.702408s","0.762392s","1.023784s","1.141627s","1.389929s","1.650156s","1.891828s","1.916019s","2.280391s","2.7421s","2.951053s","3.588713s","4.046382s","5.31069s","5.433687s","6.561518s","7.222227s","6.998665s","8.4995s","7.108052s","8.232878s","9.499977s","10.223982s","9.720852s","10.929698s","11.531332s","11.769357s","9.201112s","12.438632s","13.114624s","23.979276s","16.6972s","18.690109s"]

after: ["0.000067s","0.000155s","0.000178s","0.00023s","0.000211s","0.000263s","0.000261s","0.000309s","0.000348s","0.000498s","0.000481s","0.000639s","0.000722s","0.000505s","0.000559s","0.000582s","0.000621s","0.000645s","0.000669s","0.000682s","0.00078s","0.001144s","0.002831s","0.002931s","0.001968s","0.000998s","0.001013s","0.001149s","0.000994s","0.001084s","0.001271s","0.001158s","0.002823s","0.002984s","0.003032s","0.001605s","0.001746s","0.001566s","0.001353s","0.001442s","0.003358s","0.00439s","0.001613s","0.002042s","0.001898s","0.001713s","0.001671s","0.003743s","0.004321s","0.001794s"]

I understand this still can be dangerous. For example if someone calls (uniqueElements 3 :: Gen (NonEmpty Bool)) this will loop for ever. But I think this is close to the best we can do if we want to keep uniqueElements parametric on a.